crates/re_query_cache/src/query.rs

```diff
@@ -156,6 +133,58 @@ macro_rules! impl_query_archetype {
                 Ok(())
             };
 
+            let mut latest_at_callback = |query: &LatestAtQuery, latest_at_cache: &mut crate::LatestAtCache| {
+                re_tracing::profile_scope!("latest_at", format!("{query:?}"));
+
+                let crate::LatestAtCache { per_query_time, per_data_time } = latest_at_cache;
+
+                // Fastest path: we have an entry for this exact query, no need to look
+                // any further.
+                if let Some(query_time_bucket) = per_query_time.get(&query.at) {
+                    return iter_results(&query_time_bucket.read());
+                }
+
+                let (data_time, arch_view) = query_archetype::<A>(store, &query, entity_path)?;
+                // TODO(cmc): actual timeless caching support.
+                let data_time = data_time.unwrap_or(TimeInt::MIN);
+
+                // Fast path: we've run the query and realized that we already have the data for the resulting
+                // _data_ time, so let's use that.
+                if let Some(data_time_bucket) = per_data_time.get(&data_time) {
+                    // We now know for a fact that a query at that data time would yield the same
+                    // results: copy the bucket accordingly so that the next cache hit ends up taking the fastest path.
+                    *per_query_time.entry(data_time).or_default() = std::sync::Arc::clone(&data_time_bucket);
```

Comment:
Shouldn't we be using `query.at` here?

If not, document why not (cache invalidation?)

```suggestion
                    *per_query_time.entry(query.at).or_default() = std::sync::Arc::clone(&data_time_bucket);
```

Reply:
The comment above explains why: if `query(t=10)` yields a data time of `2` then we know that `query(t=2)` returns the same results.

But what you're saying is _also_ true: I made both co-exist.